modules/avro-exporter/src/main/scala/io/janstenpickle/trace4cats/avro/AvroSpanExporter.scala

```diff
@@ -54,66 +52,70 @@ object AvroSpanExporter {
+  def udp[F[_]: Async, G[_]: Traverse](
     host: String = agentHostname,
     port: Int = agentPort,
   ): Resource[F, SpanExporter[F, G]] = {
+    val queueCapacity = 1
+    val maxPermits = 1L
 
     def write(
       schema: Schema,
-      address: InetSocketAddress,
+      address: SocketAddress[IpAddress],
       semaphore: Semaphore[F],
       queue: Queue[F, Batch[G]],
-      socket: UDPSocket[F]
+      socket: DatagramSocket[F]
     ): F[Unit] =
       Stream
         .repeatEval {
-          (for {
-            batch <- queue.dequeue1
-            _ <- semaphore.acquire
-            _ <- batch.spans.traverse { span =>
-              for {
-                ba <- encode[F](schema)(span)
-                _ <- socket.write(Packet(address, Chunk.bytes(ba)))
-              } yield ()
+          for {
+            batch <- queue.take
+            _ <- semaphore.permit.use { _ =>
+              batch.spans.traverse { span =>
+                for {
+                  ba <- encode[F](schema)(span)
+                  _ <- socket.write(Datagram(address, Chunk.array(ba)))
+                } yield ()
+              }
             }
-          } yield ()).guarantee(semaphore.release)
+          } yield ()
         }
         .compile
         .drain
 
     for {
-      avroSchema <- Resource.liftF(AvroInstances.completedSpanSchema[F])
-      address <- Resource.liftF(Sync[F].delay(new InetSocketAddress(host, port)))
-      queue <- Resource.liftF(InspectableQueue.bounded[F, Batch[G]](1))
-      semaphore <- Resource.liftF(Semaphore[F](Long.MaxValue))
-      socketGroup <- UDPSocketGroup(blocker)
-      socket <- socketGroup.open()
-      _ <- Resource.make(
-        Stream
-          .retry(write(avroSchema, address, semaphore, queue, socket), 5.seconds, _ + 1.second, Int.MaxValue)
-          .compile
-          .drain
-          .start
-      )(fiber =>
-        Applicative[F].unit.whileM_(for {
-          queueNonEmpty <- queue.getSize.map(_ != 0)
-          semaphoreSet <- semaphore.count.map(_ == 0)
-          _ <- Timer[F].sleep(50.millis)
-        } yield queueNonEmpty || semaphoreSet) >> fiber.cancel
-      )
+      avroSchema <- Resource.eval(AvroInstances.completedSpanSchema[F])
+      host <- Resource.eval(IpAddress.fromString(host).liftTo[F](new IllegalArgumentException(s"invalid host $host")))
+      port <- Resource.eval(Port.fromInt(port).liftTo[F](new IllegalArgumentException(s"invalid port $port")))
+      address = SocketAddress(host, port)
+      queue <- Resource.eval(Queue.bounded[F, Batch[G]](queueCapacity))
+      semaphore <- Resource.eval(Semaphore[F](maxPermits))
+      socketGroup <- Network[F].datagramSocketGroup()
+      socket <- socketGroup.openDatagramSocket()
+      _ <- Resource
+        .make(
+          Stream
+            .retry(write(avroSchema, address, semaphore, queue, socket), 5.seconds, _ + 1.second, Int.MaxValue)
+            .compile
+            .drain
+            .start
+        )(fiber =>
```

Comment:
@janstenpickle 
I changed the condition for releasing the worker fiber.

The fiber may be in 2 states:
- processing a new batch - hence it has acquired a permit for it
- waiting on `queue.take` - hence it has released the permit

So if we check the available permits and it's less than the `maxPermits`, then the fiber is working on a new batch and we should sleep. Otherwise, the fiber is semantically blocked and there are no new messages in the queue.

Strange, but the previous check contained `semaphore.count.map(_ == 0)`. This is probably a mistake because `count` returns `maxPermits - <number of acquired permits>`.

Reply:
Btw, what was the intention behind using a semaphore and not a Ref? Multiple fibers working concurrently?
We can add an extra parameter and use it for the queue capacity and the number of permits/workers. And we can start n fibers. Does it make sense?

Reply:
> We can add an extra parameter and use it for the queue capacity and the number of permits/workers. And we can start n fibers. Does it make sense?

Yeah this makes sense. It was most likely a naive implementation on my part!